### DISCUSSION TITLE
How to save or update Upstream data in cutsom plugin?

### DISCUSSION THREAD

**Discussion by elricli** [apache/apisix#6200] [Q&A]
还是只能通过 HTTP API？

**Comment by tokers** [comment]
It's recommended to update APISIX upstream info through the Restful APIs, but indeed you can update these data in your own plugins, all data about a request is stored in the `ngx.ctx.api_ctx`, there are several data, you may have to get insight the source code and know how to update them, but these behaviors are error-prone, please do this only if you know what you do.

**Reply by elricli** [reply]
感谢，还有一个问题，如何在自定义插件中的 `_M.rewrite` 方法中通过 `upstream_id` 覆写上游信息？

Thanks, one more question, how do I override upstream information via 'upstream_id' in the '_M.rewrite' method in a custom plugin?

**Reply by tokers** [reply]
You may want to see [this](https://github.com/apache/apisix/blob/master/apisix/upstream.lua#L523)?

**Reply by elricli** [reply]
Thanks, But I refer to the traffic-split plugin, which looks like this:
```lua
function _M.rewrite(conf, ctx)
    -- get upstream info
    local upstreams = core.config.fetch_created_obj("/upstreams")
    if not upstreams then
        return core.response.exit(502)
    end
    -- new upstream id by system name and tag
    local upstream_id = "ups-sys1-abcd" -- dynamic upstream_id
    local ups_info = upstreams:get(tostring(upstream_id))
    if not ups_info then
        return
    end
    ups_info = ups_info.value
    upstream.set(ctx, ups_info.name, ctx.conf_version, ups_info)
end
```

**Reply by juzhiyuan** [reply]
Hi @drrrMikado, may I know what's ` 覆写上游信息？`? Do you mean we could dynamically change rule's upstream data by upstream_id?

**Reply by elricli** [reply]
> Hi @drrrMikado, may I know what's ` 覆写上游信息？`? Do you mean we could dynamically change rule's upstream data by upstream_id?

Yes, dynamically change route's upstream by custom header rule.

**Reply by juzhiyuan** [reply]
Hi @leslie-tsang, could you please give some hints?

**Reply by leslie-tsang** [reply]
Looks like we have a [canary release](https://en.wikipedia.org/wiki/Feature_toggle#Canary_release) scenario here, I will try to figure it out on the weekend.
